pad_ui/PackDownloadDiv.gd

```diff
@@ -0,0 +1,151 @@
+	return stepify(byte_value / 1000000, 0.01)
+
+func _process(delta):
+	if request_obj == null:
+		return
+
+	var should_update_ui = request_obj.get_state().get_status() != PlayAssetPackManager.AssetPackStatus.UNKNOWN
+	
+	if should_update_ui:
+		# update the downloading UI
+		var bytes_downloaded = float(request_obj.get_state().get_bytes_downloaded())
+		var bytes_to_download = float(request_obj.get_state().get_total_bytes_to_download())
+		var progress_ratio = bytes_downloaded / bytes_to_download
+		var downloaded_megabyte = _convert_byte_to_stepified_megabytes(bytes_downloaded)
+		var total_megabyte = _convert_byte_to_stepified_megabytes(bytes_to_download)
+		
+		if request_obj.get_state().get_status() != PlayAssetPackManager.AssetPackStatus.COMPLETED:
+			var progress_text_format_string = "%fMB/%fMB"
+			$ProgressText.text = progress_text_format_string % [downloaded_megabyte, total_megabyte]
+			# start progress bar animation
+			var tween_duration = 0.4
+			$Tween.interpolate_property($ProgressBar, "value", $ProgressBar.value, \
+				int(progress_ratio*100), tween_duration, Tween.TRANS_QUART, Tween.EASE_OUT)
+			$Tween.start()
+
+func _reset_download_ui():
+	$ProgressBar.value = 0
+	$ProgressText.text = "-MB/-MB"
+	
+func _on_cancel_button_pressed():
+	_hide_downloading_ui()
+	_show_init_ui()
+	pad_manager.cancel_asset_pack_request(pack_name)
+
+func _on_retry_button_pressed():
+	# clear previous download value
+	request_obj = pad_manager.fetch_asset_pack(pack_name)
+	request_obj.connect("request_completed", self, "_handle_request_completed")
+	_reset_download_ui()
+	_hide_failed_ui()
+	_show_downloading_ui()
+
+func _on_download_button_pressed():
+	# clear previous download value
+	request_obj = pad_manager.fetch_asset_pack(pack_name)
+	request_obj.connect("request_completed", self, "_handle_request_completed")
+	_reset_download_ui()
+	_hide_init_ui()
+	_show_downloading_ui()
+
+func _on_delete_button_pressed():
+	_hide_completed_ui()
+	_show_init_ui()
+	pad_manager.remove_pack(pack_name)
+	# update downloaded_packs var in Main.gd
+	emit_signal("removed_pack", pack_name)
+
+func _show_init_ui():
+	$TerminalStateText.text = "Not Installed"
+	$TerminalStateText.show()
+	$DownloadButton.show()
+
+func _hide_init_ui():
+	$TerminalStateText.hide()
+	$DownloadButton.hide()
+
+func _show_downloading_ui():
+	$ProgressBar.show()
+	$ProgressText.show()
+	$CancelButton.show()
+
+func _hide_downloading_ui():
+	$ProgressBar.hide()
+	$ProgressText.hide()
+	$CancelButton.hide()
+
+func _show_completed_ui():
+	# added extra spaces to account for the checkmark icon
+	$TerminalStateText.text = "    Download Completed"
+	$TerminalStateText.show()
+	$CompletedIcon.show()
+	$DeleteButton.show()
+
+func _hide_completed_ui():
+	$CompletedIcon.hide()
+	$TerminalStateText.hide()
+	$DeleteButton.hide()
+
+func _show_failed_ui():
+	# added extra spaces to account for the error icon
+	$TerminalStateText.text = "    Failed"
```

Comment:
Is it easy to make that "    " a constant and reuse where needed? Then the explanatory comment only needs to be on the constant instead of at each usage.

Reply:
Yeah that makes sense. I refactored this "  " to ```TERMINAL_STATE_TEXT_PADDING``` constant and added more related comments